pkg/resource/node-resource-manager.go

```diff
@@ -0,0 +1,233 @@
+			return
+		}
+
+		o.BuildNodeStatus(tsp, nodeCopy)
+		if !equality.Semantic.DeepEqual(&node.Status, &nodeCopy.Status) {
+			// Update Node status extend-resource info
+			// TODO fix: strategic merge patch kubernetes
+			if _, err := o.client.CoreV1().Nodes().Update(context.TODO(), nodeCopy, metav1.UpdateOptions{}); err != nil {
+				klog.Errorf("Failed to update node %s's status extend-resource, %v", nodeCopy.Name, err)
+				return
+			}
+			klog.V(4).Infof("Update Node %s Extend Resource Success according to TSP %s", node.Name, tsp.Name)
+		}
+		return
+	}
+}
+
+func (o *NodeResourceManager) getNode() (*v1.Node, []v1.NodeAddress) {
+	node, err := o.nodeLister.Get(o.nodeName)
+	if err != nil {
+		klog.Errorf("Failed to get node: %v", err)
+		return nil, []v1.NodeAddress{}
+	}
+	return node, node.Status.Addresses
+}
+
+func (o *NodeResourceManager) FindTargetNode(tsp *predictionapi.TimeSeriesPrediction, addresses []v1.NodeAddress) (bool, error) {
+	address := tsp.Spec.TargetRef.Name
+	if address == "" {
+		return false, fmt.Errorf("Tsp %s target is not specified", tsp.Name)
+	}
+
+	// the reason we use node ip instead of node name as the target name is
+	// some monitoring system does not persist node name
+	for _, addr := range addresses {
+		if addr.Address == address {
+			return true, nil
+		}
+	}
+	klog.V(4).Infof("Target %s mismatch this node", address)
+	return false, nil
+}
+
+func (o *NodeResourceManager) BuildNodeStatus(tsp *predictionapi.TimeSeriesPrediction, node *v1.Node) {
+	idToResourceMap := map[string]*v1.ResourceName{}
+	for _, metrics := range tsp.Spec.PredictionMetrics {
+		if metrics.ResourceQuery == nil {
+			continue
+		}
+		idToResourceMap[metrics.ResourceIdentifier] = metrics.ResourceQuery
+	}
+	// build node status
+	nextPredictionResourceStatus := &tsp.Status
+	for _, metrics := range nextPredictionResourceStatus.PredictionMetrics {
+		resourceName, exists := idToResourceMap[metrics.ResourceIdentifier]
+		if !exists {
+			continue
+		}
+		for _, timeSeries := range metrics.Prediction {
+			var maxUsage, nextUsage float64
+			var nextUsageFloat float64
+			var err error
+			for _, sample := range timeSeries.Samples {
+				if nextUsageFloat, err = strconv.ParseFloat(sample.Value, 64); err != nil {
+					klog.Errorf("Failed to parse extend resource value %v: %v", sample.Value, err)
+					continue
+				}
+				nextUsage = nextUsageFloat
+				if maxUsage < nextUsage {
+					maxUsage = nextUsage
+				}
+			}
+			var nextRecommendation float64
+			switch *resourceName {
+			case v1.ResourceCPU:
+				// cpu need to be scaled to m as ext resource cannot be decimal
+				nextRecommendation = (float64(node.Status.Allocatable.Cpu().Value()) - maxUsage) * 1000
+			case v1.ResourceMemory:
+				// unit of memory in prometheus is in Ki, need to be converted to byte
+				nextRecommendation = float64(node.Status.Allocatable.Memory().Value()) - (maxUsage * 1000)
+			default:
+				continue
+			}
+			if nextRecommendation <= 0 {
```

Comment:
If the recommendation is zero, set the extend resource to zero is more better ?

Reply:
OK, i correct it to <

Reply:
If the recommendation is less than zero, the extension resource should also be set to zero, right? In the presence of a watermark, the recommendation is likely to be less than 0